### PR TITLE
Initial recursiveStrict implementation

### PR DESCRIPTION
**Pull request by iansu**
##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: lib/internal/fs/rimraf.js [hunk 1]
```diff
@@ -42,6 +42,34 @@ const separator = Buffer.from(sep);
...
+    if (stat(path, (err, stats) => {
+      if (err) {
+        if (err.code === 'ENOENT') {
+          throw err;
+        } else if (!stats.isDirectory()) {
+          throw new ERR_INVALID_ARG_VALUE('path', path, 'is not a directory')
+        }
+      }
+
+      _rimraf(path, options, function CB(err) {
```
Comment (iansu) [inline lib/internal/fs/rimraf.js]: Repeating this function seems bad. I could move it into an external function or something else?

File: lib/internal/fs/rimraf.js [hunk 2]
```diff
@@ -42,6 +42,34 @@ const separator = Buffer.from(sep);
 function rimraf(path, options, callback) {
   let retries = 0;
 
+  if (options.recursiveStrict) {
+    if (stat(path, (err, stats) => {
```
Comment (iansu) [inline lib/internal/fs/rimraf.js]: I think we need `stat` vs `lstat` because we actually want to follow the symlink but I'm not 100% sure.

Comment (bcoe) [reply lib/internal/fs/rimraf.js]: I would check the behavior of `rmdir` on a symlink, without recursive, and just match it.

File: lib/internal/fs/rimraf.js (hunk 1 above)
Comment (bcoe) [reply lib/internal/fs/rimraf.js]: Yeah, I wonder if you could just pass `strict` down into the other helper, and vary the behavior accordingly?

**Review by bcoe — COMMENTED** [review]
I think the most important thing for this PR will be fleshing out the tests in `test/parallel/test-fs-rmdir-recursive.js`, for every case except deleting a nested folder, I think we should match the errors returned by `rmdir` with recursive `false`.

**Comment by iansu** [comment]
@bcoe Agreed. I've already started on that.

File: test/parallel/test-fs-rmdir-recursive.js [hunk 3]
```diff
@@ -105,6 +100,25 @@ function removeAsync(dir) {
...
+    common.mustCall((err) => {
+      assert.strictEqual(err.code, 'ENOENT');
+    })
+  );
+
+  // Should fail if target is a file
+  const filePath = path.join(tmpdir.path, 'rmdir-async-file.txt');
+  fs.writeFileSync(filePath, '');
+  fs.rmdir(filePath, { recursive: true }, common.mustCall((err) => {
+    assert.strictEqual(err.code, 'ERR_INVALID_ARG_VALUE');
```
Comment (bcoe) [inline test/parallel/test-fs-rmdir-recursive.js]: I'd be tempted to make this an ENODIR? or whatever the equivalent is for `rmdir` when it's not recursive.

File: lib/internal/fs/rimraf.js [hunk 4]
```diff
@@ -39,6 +44,36 @@ const readdirEncoding = 'buffer';
...
+          'path', path, 'is not a directory'
+        ));
+      } else {
+        callback();
+      }
+    });
+  } else {
+    let stats;
+
+    try {
```
Comment (bcoe) [inline lib/internal/fs/rimraf.js]: curious what the reasoning is for suppressing other types of errors?

**Pull request closed without merging**